### ISSUE TITLE
Recursively nested properties and compaction

### ISSUE THREAD

**Issue by pchampin** [w3c/json-ld-api#391]
Following the discussion in #380, I have a question:

Consider the following data:
```json
{
  "@context": {
    "foo": "http://example.org/foo",
    "bar": "http://example.org/bar",
    "baz": "http://example.org/baz",
    "n1": "@nest",
    "n2": "@nest"
  },
  "foo": "FOO",
  "n1": {
    "bar": "BAR",
    "n2": {
      "baz": "BAZ"
    }
  }
}
```

Can I expand it and compact it back to the *same* shape (including the doubly-nested property `baz`)?

I gave it a try [there](https://json-ld.org/playground/#startTab=tab-compacted&json-ld=%7B%22%40context%22%3A%7B%22foo%22%3A%22http%3A%2F%2Fexample.org%2Ffoo%22%2C%22bar%22%3A%22http%3A%2F%2Fexample.org%2Fbar%22%2C%22baz%22%3A%22http%3A%2F%2Fexample.org%2Fbaz%22%2C%22n1%22%3A%22%40nest%22%2C%22n2%22%3A%22%40nest%22%7D%2C%22foo%22%3A%22FOO%22%2C%22n1%22%3A%7B%22bar%22%3A%22BAR%22%2C%22n2%22%3A%7B%22baz%22%3A%22BAZ%22%7D%7D%7D&context=%7B%22%40context%22%3A%7B%22foo%22%3A%7B%22%40id%22%3A%22http%3A%2F%2Fexample.org%2Ffoo%22%7D%2C%22bar%22%3A%7B%22%40id%22%3A%22http%3A%2F%2Fexample.org%2Fbar%22%2C%22%40nest%22%3A%22n1%22%7D%2C%22baz%22%3A%7B%22%40id%22%3A%22http%3A%2F%2Fexample.org%2Fbaz%22%2C%22%40nest%22%3A%22n2%22%7D%2C%22n1%22%3A%7B%22%40id%22%3A%22%40nest%22%7D%2C%22n2%22%3A%7B%22%40id%22%3A%22%40nest%22%2C%22%40nest%22%3A%22n1%22%7D%7D%7D) but with no luck.

If this is indeed impossible, I'm **not** suggesting we should fix it now. But I'm trying to decide how bad it is if we introduce an asymmetry between expansion and compaction wrt nested properties in #388. If it appears that the asymetry is already there (in the example above), that would help be byte the bullet.

**Comment by niklasl** [comment]
I attempted a similar thing, and I think it could, in theory, only work by declaring one of the nests as nested within the other, e.g.:
```json
    "n1": "@nest",
    "n2": {"@id": "@nest", "@nest": "n1"}
```
But the compaction algorithm doesn't support that.

**Comment by gkellogg** [comment]
No, one level only. Basically, when compacting, we match to the term "baz" and see that it is nested on "n2". There is nothing to say that "n2" is, itself, nested under "n1".

In theory, (but I don't believe in practice), you might use the following context:

```json
{
  "@context": {
    "foo": "http://example.org/foo",
    "bar": {"@id": "http://example.org/bar", "@nest": "n1"},
    "baz": {"@id": "http://example.org/baz", "@nest": "n2"},
    "n1": "@nest",
    "n2": {"@id": "@nest", "@nest": "n1"}
  },
  "foo": "FOO",
  "n1": {
    "bar": "BAR",
    "n2": {
      "baz": "BAZ"
    }
  }
}
```

But, the processing rules don't allow that.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2020/2020-02-28-json-ld#section3-2).

- `RESOLVED:  Defer #391 on recursively nested properties to future version as new feature that can be experimented with`
<details><summary><i class="differentiate">View the transcript</i></summary>
Recursive Nesting and Compaction<br/>
Rob Sanderson: <a href="https://github.com/w3c/json-ld-api/issues/391">https://github.com/w3c/json-ld-api/issues/391</a><br/>
Rob Sanderson: Next one is related by pchampin.<br/>
Pierre-Antoine Champin: This issue is just to check that recursively nested properties are … after round-tripping.<br/>
… I didn’t suggest any change. This can be closed. This was just to support my other arguments in the previous issue.<br/>
Gregg Kellogg: We should also defer this, rather than close.<br/>
Proposed resolution: Defer #391 on recursively nested properties to future version as new feature that can be experimented with (Rob Sanderson)<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
Pierre-Antoine Champin: +1<br/>
David I. Lehn: +1<br/>
Ruben Taelman: +1<br/>
Tim Cole: +1<br/>
Harold Solbrig: +1<br/>
Resolution #2: Defer #391 on recursively nested properties to future version as new feature that can be experimented with<br/>
3.<br/>
</details>

**Comment by MajdT51** [comment]
Hello, What could I do if I must have Recursively nested properties in my json-ld, is there a workaround or a trick :)? Thanks
```
{
   "data": {
            "value": "TestValue",
            "nameType": {
                "testKey": "testValue",
                "name": "TestName"
            }
        }
}
```

**Comment by MajdT51** [comment]
> @MajdT51 without a context in your data, and a description of what you are trying to achieve with it, it is hard to understand exactly what you question is...

Thanks @pchampin , I have a recursive nested object "nameType" inside object "data". And I'm using this context 
```
{
  "@context": {
    "ctx": "https://mydomain/selfdescription#",
    "@version": 1.1,
    "@protected": true,
    "NameCredential": {
      "@id": "ctx:NameCredential",
      "@context": {
        "data": "@nest",
        "value": {
          "@id": "ctx:value",
          "@nest": "data"
        },
        "nameType": {
          "@id": "ctx:nameType",
          "@nest": "data",
          "@context": {
            "name": {
              "@id": "ctx:name",
              "@nest": "nameType"
            },
            "testKey": {
              "@id": "ctx:testKey",
              "@nest": "nameType"
            }
          }
        }
      }
    }
  }
}
```

However it is not valid and pyld shows me this error 
JSON-LD compact error; nested property must have an @nest value resolving to @nest.( https://github.com/digitalbazaar/pyld/blob/master/lib/pyld/jsonld.py#L3353) because nameType must be also like "data" be @nest

**Comment by pchampin** [comment]
the two lines `"@nest": "nameType"` is not necessary in your context. As part of a scoped context, those properties (`name` and `testKey`) will only be used on values of `nameType`, so they will be "nested" the way you want.
